# Exercise

In [139]:
# Importing libraires 
import pandas as pd
import numpy as np
import os

In [140]:
path = r'C:\Users\mauri\OneDrive\Documentos\Instacart Basket Analysis\02 Data'

In [141]:
df_ords = pd.read_csv(os.path.join(path, 'Original Data', 'orders.csv'), index_col = False)

In [142]:
df_prods = pd.read_csv(os.path.join(path, 'Original Data', 'products.csv'), index_col = False)

In [143]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [144]:
df_ords_2 = df_ords.drop(columns = ['eval_set'])

In [145]:
df_ords['days_since_prior_order'].value_counts(dropna = False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [146]:
# Renaming order_dow
df_ords.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [147]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [148]:
# changing df variable
df_ords['order_id'] = df_ords['order_id'].astype('str')

In [149]:
df_ords['order_id'].dtype

dtype('O')

In [150]:
# Importing df dep
df_dep = pd.read_csv(os.path.join(path, 'Original Data', 'departments.csv'))

In [151]:
df_dep.head()

department_id       1      2       3        4        5              6  \
0    department  frozen  other  bakery  produce  alcohol  international   

           7     8                9  ...            12      13         14  \
0  beverages  pets  dry goods pasta  ...  meat seafood  pantry  breakfast   

             15          16         17      18      19    20       21  
0  canned goods  dairy eggs  household  babies  snacks  deli  missing  

[1 rows x 22 columns]

In [152]:
#df transposing 
df_dep.T

0
department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [153]:
df_dep_t = df_dep.T

In [154]:
# applying new index
df_dep_t.reset_index()

index                0
0   department_id       department
1               1           frozen
2               2            other
3               3           bakery
4               4          produce
5               5          alcohol
6               6    international
7               7        beverages
8               8             pets
9               9  dry goods pasta
10             10             bulk
11             11    personal care
12             12     meat seafood
13             13           pantry
14             14        breakfast
15             15     canned goods
16             16       dairy eggs
17             17        household
18             18           babies
19             19           snacks
20             20             deli
21             21          missing

In [155]:
#copy of first row
new_header = df_dep_t.iloc[0]

In [156]:
new_header

0    department
Name: department_id, dtype: object

In [157]:
# copy of df without first row
df_dep_t_new = df_dep_t[1:]

In [158]:
df_dep_t_new

0
1            frozen
2             other
3            bakery
4           produce
5           alcohol
6     international
7         beverages
8              pets
9   dry goods pasta
10             bulk
11    personal care
12     meat seafood
13           pantry
14        breakfast
15     canned goods
16       dairy eggs
17        household
18           babies
19           snacks
20             deli
21          missing

In [159]:
# add new headers to df
df_dep_t_new.columns = new_header

In [160]:
df_dep_t_new

department_id       department
1                       frozen
2                        other
3                       bakery
4                      produce
5                      alcohol
6                international
7                    beverages
8                         pets
9              dry goods pasta
10                        bulk
11               personal care
12                meat seafood
13                      pantry
14                   breakfast
15                canned goods
16                  dairy eggs
17                   household
18                      babies
19                      snacks
20                        deli
21                     missing

In [161]:
data_dict = df_dep_t_new.to_dict('index')

In [162]:
data_dict

{'1': {'department': 'frozen'},
 '2': {'department': 'other'},
 '3': {'department': 'bakery'},
 '4': {'department': 'produce'},
 '5': {'department': 'alcohol'},
 '6': {'department': 'international'},
 '7': {'department': 'beverages'},
 '8': {'department': 'pets'},
 '9': {'department': 'dry goods pasta'},
 '10': {'department': 'bulk'},
 '11': {'department': 'personal care'},
 '12': {'department': 'meat seafood'},
 '13': {'department': 'pantry'},
 '14': {'department': 'breakfast'},
 '15': {'department': 'canned goods'},
 '16': {'department': 'dairy eggs'},
 '17': {'department': 'household'},
 '18': {'department': 'babies'},
 '19': {'department': 'snacks'},
 '20': {'department': 'deli'},
 '21': {'department': 'missing'}}

In [163]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [164]:
print(data_dict.get('19'))

{'department': 'snacks'}


In [165]:
# Filtering data 1
df_snacks =  df_prods[df_prods['department_id']==19]

In [166]:
df_snacks

product_id                                   product_name  aisle_id  \
0               1                     Chocolate Sandwich Cookies        61   
15             16                  Mint Chocolate Flavored Syrup       103   
24             25        Salted Caramel Lean Protein & Fiber Bar         3   
31             32                  Nacho Cheese White Bean Chips       107   
40             41    Organic Sourdough Einkorn Crackers Rosemary        78   
...           ...                                            ...       ...   
49666       49662                   Bacon Cheddar Pretzel Pieces       107   
49669       49665  Super Dark Coconut Ash & Banana Chocolate Bar        45   
49670       49666                  Ginger Snaps Snacking Cookies        61   
49675       49671                           Milk Chocolate Drops        45   
49683       49679                        Famous Chocolate Wafers        61   

       department_id  prices  
0                 19     5.8  
15                19     5.2  
24                19     1.9  
31                19     4.9  
40                19     6.5  
...              ...     ...  
49666             19     3.6  
49669             19     6.9  
49670             19     5.2  
49675             19     3.0  
49683             19     6.0  

[6264 rows x 5 columns]

In [167]:
# Filtering data 2
df_snacks_2 = df_prods.loc[df_prods['department_id'] == 19]

In [168]:
# Filtering data 3 (multiple value accepted)
df_snacks_3 = df_prods.loc[df_prods['department_id'].isin([19])]

# Task 

In [169]:
df_ords.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0  2539329        1    prior             1                   2   
1  2398795        1    prior             2                   3   
2   473747        1    prior             3                   3   
3  2254736        1    prior             4                   4   
4   431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [170]:
#step 2
df_ords['user_id'] = df_ords['user_id'].astype('str')

In [171]:
#step 3
df_ords.rename(columns = {'days_since_prior_order' : 'days_since_last_order'}, inplace = True)

#### Step 4
##### Your client wants to know what the busiest hour is for placing orders. Find the frequency of the corresponding variable and share your findings.

##### A/ 10 is the most frecuent hour

In [172]:
df_ords.mode()

order_id user_id eval_set  order_number  orders_day_of_week  \
0              1  100142    prior           1.0                 0.0   
1             10  100316      NaN           2.0                 NaN   
2            100  100325      NaN           3.0                 NaN   
3           1000   10046      NaN           4.0                 NaN   
4          10000  100514      NaN           NaN                 NaN   
...          ...     ...      ...           ...                 ...   
3421078   999995     NaN      NaN           NaN                 NaN   
3421079   999996     NaN      NaN           NaN                 NaN   
3421080   999997     NaN      NaN           NaN                 NaN   
3421081   999998     NaN      NaN           NaN                 NaN   
3421082   999999     NaN      NaN           NaN                 NaN   

         order_hour_of_day  days_since_last_order  
0                     10.0                   30.0  
1                      NaN                    NaN  
2                      NaN                    NaN  
3                      NaN                    NaN  
4                      NaN                    NaN  
...                    ...                    ...  
3421078                NaN                    NaN  
3421079                NaN                    NaN  
3421080                NaN                    NaN  
3421081                NaN                    NaN  
3421082                NaN                    NaN  

[3421083 rows x 7 columns]

In [173]:
#step 5
data_dict.get('4')

{'department': 'produce'}

In [174]:
#step 6
df_breakfast = df_prods[df_prods['department_id'] == 14]

In [175]:
df_breakfast

product_id                                       product_name  \
27             28                                  Wheat Chex Cereal   
33             34                                                NaN   
67             68                            Pancake Mix, Buttermilk   
89             90                                       Smorz Cereal   
210           211   Gluten Free Organic Cereal Coconut Maple Vanilla   
...           ...                                                ...   
49330       49326                            Cereal Variety Fun Pack   
49395       49391            Light and Fluffy Buttermilk Pancake Mix   
49547       49543                          Chocolate Cheerios Cereal   
49637       49633               Shake 'N Pour Buttermilk Pancake Mix   
49667       49663  Ultra Protein Power Crunch Peanut Butter N' Ho...   

       aisle_id  department_id  prices  
27          121             14    10.1  
33          121             14    12.2  
67          130             14    13.7  
89          121             14     3.9  
210         130             14     3.6  
...         ...            ...     ...  
49330       121             14     9.1  
49395       130             14     2.0  
49547       121             14    10.8  
49637       130             14    14.2  
49667        57             14    10.2  

[1116 rows x 5 columns]

In [176]:
# sterp 7
df_dinner = df_prods.loc[df_prods['department_id'].isin([5, 20, 7, 12])] 

In [177]:
df_dinner

product_id                                    product_name  aisle_id  \
2               3            Robust Golden Unsweetened Oolong Tea        94   
6               7                  Pure Coconut Water With Orange        98   
9              10  Sparkling Orange Juice & Prickly Pear Beverage       115   
10             11                               Peach Mango Juice        31   
16             17                               Rendered Duck Fat        35   
...           ...                                             ...       ...   
49676       49672                          Cafe Mocha K-Cup Packs        26   
49679       49675             Cinnamon Dolce Keurig Brewed K Cups        26   
49680       49676                          Ultra Red Energy Drink        64   
49686       49682                              California Limeade        98   
49688       49684       Vodka, Triple Distilled, Twist of Vanilla       124   

       department_id  prices  
2                  7     4.5  
6                  7     4.4  
9                  7     8.4  
10                 7     2.8  
16                12    17.1  
...              ...     ...  
49676              7     6.5  
49679              7    14.0  
49680              7    14.5  
49686              7     4.3  
49688              5     5.3  

[7650 rows x 5 columns]

#### Step 8
##### It’s important that you keep track of total counts in your dataframes. How many rows does the last dataframe you created have? 
##### A/ 7650

In [178]:
df_dinner.shape

(7650, 5)

In [179]:
#Step 9
df_user1 = df_ords[df_ords['user_id'] == 1]

In [180]:
df_user1

Empty DataFrame
Columns: [order_id, user_id, eval_set, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order]
Index: []

In [181]:
# Step 10
df_user1.describe()

order_number  orders_day_of_week  order_hour_of_day  \
count           0.0                 0.0                0.0   
mean            NaN                 NaN                NaN   
std             NaN                 NaN                NaN   
min             NaN                 NaN                NaN   
25%             NaN                 NaN                NaN   
50%             NaN                 NaN                NaN   
75%             NaN                 NaN                NaN   
max             NaN                 NaN                NaN   

       days_since_last_order  
count                    0.0  
mean                     NaN  
std                      NaN  
min                      NaN  
25%                      NaN  
50%                      NaN  
75%                      NaN  
max                      NaN

In [182]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_wrangled.csv'))

In [183]:
df_dep_t_new.to_csv(os.path.join(path, 'Prepared Data','departments_wrangled.csv'))